src/spark_rapids_ml/knn.py

```diff
@@ -493,3 +521,100 @@ def _get_cuml_transform_func(
+        self,
+        datasetA: DataFrame,
+        datasetB: DataFrame,
+        numNearestNeighbors: int,
+        distCol: str = "distCol",
+    ) -> DataFrame:
+        """
+        This function returns the k exact nearest neighbors (knn) in datasetA regarding each query vector in datasetB.
+        The purpose is to match pyspark LSH approxSimilarityJoin(datasetA, datasetB, threshold, distCol). Note this function
+        is directional that treats datasetA as the item dataframe and datasetB as the query dataframe. This is because
+        knn-based search requires the role of query vectors, which is different from threshold-based search where datasetA and
+        datasetB are interchangeable.
+
+        Parameters
+        ----------
+        datasetA: pyspark.sql.DataFrame
+            the item_df dataframe from which the k nearest neighbors will be retrieved. Each row represents an item vector.
+
+        datasetB: pyspark.sql.DataFrame
+            the query_df dataframe. Each row represents a query vector.
+
+        numNearestNeighbors: int
+            the number of exact nearest neighbors to be returned for each query.
+
+        distCol: str
+            the name of the distance column
+
+        Returns
+        -------
+        knnjoin_df: pyspark.sql.DataFrame
+            the result dataframe that has three columns (datasetA, datasetB, distCol).
+            datasetA column is of struct type that concatenates all the columns of input dataframe datasetA.
+            Similarly, datasetB column is of struct type that concatenates all the columns of input dataframe datasetB.
+            distCol is the distance column. A row in knnjoin_df is in the format (v1, v2, dist(v1, v2)),
+            where item_vector v1 is one of the k nearest neighbors of query_vector v2 and their distance is dist(v1, v2).
+        """
+
+        # save relevant self variables
+        estimator_item_df_withid = self._item_df_withid
+        estimator_processed_item_df = self._processed_item_df
+        estimator_n_neighbors = self.cuml_params["n_neighbors"]
+
+        # modify relevant self variables
+        id_col_name = self.getIdCol()
+
+        self._item_df_withid = self.ensureIdCol(datasetA)
+
+        self._processed_item_df = self._item_df_withid.withColumn(
+            alias.label, lit(self._label_isdata)
+        )
+        self.cuml_params["n_neighbors"] = numNearestNeighbors
+
+        # call kneighbors then prepare return results
+        (item_df_withid, query_df_withid, knn_df) = self.kneighbors(datasetB)
+
+        from pyspark.sql.functions import arrays_zip, col, explode, struct
+
+        knn_pair_df = knn_df.select(
+            f"query_{id_col_name}",
+            explode(arrays_zip("indices", "distances")).alias("zipped"),
+        ).select(
+            f"query_{id_col_name}",
+            col("zipped.indices").alias(f"item_{id_col_name}"),
+            col("zipped.distances").alias(distCol),
+        )
```

Comment:
In a follow up, should we add an option to filter out pairs whose distance is > some threshold?

Reply:
Created a jira ticket to track this. 

Specifically, do you mean to add an approxSimilarityJoin function or add an argument to the existing exactNearestNeighborsJoin function?